_posts/2020-05-18-websocket-fallback-sockjs.md

```diff
@@ -0,0 +1,329 @@
+{% endhighlight %}
+
+WebSocketConfig.java
+{% highlight diff %}
+@Configuration
+@EnableWebSocket
++@EnableWebMvc
+-public class WebSocketConfig implements WebSocketConfigurer {
++public class WebSocketConfig implements WebSocketConfigurer, WebMvcConfigurer {
+    private SocketHandler socketHandler;
+
+    public WebSocketConfig(SocketHandler socketHandler) {
+        this.socketHandler = socketHandler;
+    }
+
+    @Override
+    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
+-       registry.addHandler(socketHandler, "/signaling");
++       registry.addHandler(socketHandler, "/signaling"); // pozostawione dla kompatybilności wstecz :)
++       registry.addHandler(socketHandler, "/v2/signaling/sockjs-server") // endpoint dla SockJS
++               .withSockJS()
++               // SockJS wykorzystuje iframe w specyficznych przypadkach
++               // i tutaj "nadmiarowo" podajemy własną wersję biblioteki JS z własnego serwera,
++               // zamiast https://cdn.jsdelivr.net/sockjs/1.0.0/sockjs.min.js
++               .setClientLibraryUrl("/v2/signaling/sockjs-static/sockjs-client/sockjs.min.js");
+    }
++   @Override
++   public void addResourceHandlers(ResourceHandlerRegistry registry) {
++       // tutaj również nadmiarowo wystawiamy statyczny zasób z biblioteką SockJS
++       registry.addResourceHandler("/v2/signaling/sockjs-static/sockjs-client/*.min.js")
++               .addResourceLocations("classpath:/META-INF/resources/webjars/sockjs-client/1.4.0/dist/")
++               .setCachePeriod(3600)
++               .resourceChain(true)
++               .addResolver(new PathResourceResolver());
++    }
+}
+{% endhighlight %}
+
+#### Frontend
+web-socket-manager.ts
+{% highlight diff %}
+import {SockjsScriptLoader} from 'websocket/sockjs-script-loader';
+ 
+export class WebSocketManager {
+   private webSocket?: WebSocket;
+   constructor() {
+-       this.webSocket = this.initWebSocket();
++       new SockjsScriptLoader().loadSockJs()
++               .then(() => {
++                   this.webSocket = this.initWebSocket();
++               })
++               .catch(error => {
++                   this.webSocket = this.initWebSocket();
++               });
+   }
+
+    initWebSocket(): WebSocket {
++       // @ts-ignore
++       if (typeof SockJS === 'function') { // objeście problemu z RollUp ;)
++           // @ts-ignore
++           return new SockJS('https://localhost/v2/signaling/sockjs-server');
++       } else {
+-           return new WebSocket('https://localhost/signaling');
++           return new WebSocket('https://localhost/v2/signaling/sockjs-server/websocket');
++       }
+    }
+}
+{% endhighlight %}
+
+
+
+{% highlight typescript %}
+export class SockjsScriptLoader {
+    private sockJsScript = '/v2/signaling/sockjs-static/sockjs-client/sockjs.min.js';
+    private state: ('loaded'|'loading'|'unloaded'|'error') = 'unloaded';
+
+    public loadSockJs(): Promise<void> {
```

Comment:
bardzo prowizoryczna ta implementacja :(
jak już wrzucamy to:
a) zróbmy to dobrze
b) użyjmy liba

a w ogóle to chyba nie jest potrzebne bo przewija się ciągle temat rollupa którego tu nie powinno być